text/0000-inline-asm.md

```diff
@@ -0,0 +1,918 @@
+reg_spec := <register class> / "<explicit register>"
+operand_expr := expr / "_" / expr "=>" expr / expr "=>" "_"
+reg_operand := dir_spec "(" reg_spec ")" operand_expr
+operand := reg_operand / "const" const_expr / "sym" path
+option := "pure" / "nomem" / "readonly" / "preserves_flags" / "noreturn"
+options := "options(" option *["," option] ")"
+asm := "asm!(" format_string *("," [ident "="] operand) ["," options] ")"
+```
+
+[format-syntax]: https://doc.rust-lang.org/std/fmt/#syntax
+
+## Template string
+
+The assembler template uses the same syntax as [format strings][format-syntax] (i.e. placeholders are specified by curly braces). The corresponding arguments are accessed in order, by index, or by name. However, implicit named arguments (introduced by [RFC #2795][rfc-2795]) are not supported.
+
+The assembly code syntax used is that of the GNU assembler (GAS). The only exception is on x86 where the Intel syntax is used instead of GCC's AT&T syntax.
+
+This RFC only specifies how operands are substituted into the template string. Actual interpretation of the final asm string is left to the assembler.
+
+However there is one restriction on the asm string: any assembler state (e.g. the current section which can be changed with `.section`) must be restored to its original value at the end of the asm string.
+
+The compiler will lint against any operands that are not used in the template string, except for operands that specify an explicit register.
+
+[rfc-2795]: https://github.com/rust-lang/rfcs/pull/2795
+
+## Operand type
+
+Several types of operands are supported:
+
+* `in(<reg>) <expr>`
+  - `<reg>` can refer to a register class or an explicit register. The allocated register name is substituted into the asm template string.
+  - The allocated register will contain the value of `<expr>` at the start of the asm code.
+  - The allocated register must contain the same value at the end of the asm code (except if a `lateout` is allocated to the same register).
+* `out(<reg>) <expr>`
+  - `<reg>` can refer to a register class or an explicit register. The allocated register name is substituted into the asm template string.
+  - The allocated register will contain an unknown value at the start of the asm code.
+  - `<expr>` must be a (possibly uninitialized) place expression, to which the contents of the allocated register is written to at the end of the asm code.
+  - An underscore (`_`) may be specified instead of an expression, which will cause the contents of the register to be discarded at the end of the asm code (effectively acting as a clobber).
+* `lateout(<reg>) <expr>`
+  - Identical to `out` except that the register allocator can reuse a register allocated to an `in`.
+  - You should only write to the register after all inputs are read, otherwise you may clobber an input.
+  - `lateout` must be used instead of `out` if you are specifying the same explicit register as an `in`.
```

Comment:
This has the same semantic meaning as having `inlateout` with the same register. I.e. `in("eax") <in_expr>, lateout("eax") <out_expr>` can be rewritten as `inlateout("eax") <in_expr> => <out_expr>`.

This should be at least noted here, but perhaps completely forbidden? There should be preferably one way to do a thing, and this rounds it out, mandating that a specific register can only participate in one directive. It also makes the code more readable, when you immediately see that the register is used for more than one purpose.

Reply:
This is a bit tricky. The reason why I decided to allow this is that in the case of `in(reg) <in_expr>, lateout(reg) <out_expr>` it is perfectly valid for the compiler to allocate the same register for both operands. In that spirit, it makes sense for `in("eax") <in_expr>, lateout("eax") <out_expr>` to work as well.

However you are right that this leaves us with two ways of doing the same thing, which could be confusing. I am not sure which option is the right way to go.

Reply:
I think the difference, is that in the `reg` example, while you can't assume that they are not the same, you also can't assume that they are. Written out (contrived example): In
```
asm!("
mov {1}, {0}
add {1}, {0}
", in(reg) x, lateout(reg) y)
```
the `mov` is required; whereas in
```
asm!("
mov {1}, {0}
add {1}, {0}
", in("eax") x, lateout("eax") y)
```
It really isn't, but this isn't immediately clear. Refactoring,
```
asm!("
mov {0}, {0}
add {0}, {0}
", inlateout("eax") x => y)
```
make it immediately clear that it's redundant, and can therefore be removed.

Reply:
In your first example the `mov` is wrong regardless since it may clobber the input value if the same register is allocated to both operands.

Reply:
> In your first example the `mov` is wrong regardless since it may clobber the input value if the same register is allocated to both operands.

not really, the first example does work when both `{0}` and `{1}` are the same register, as is demonstrated by the later example using `eax`. admittedly, this is somewhat of a special case where the input is overwritten by the same value.

Reply:
Or we could change the example to
```
push {0}
pop {1}
```
if needed. The point is, that whereas `in(reg), lateout(reg)` and `inlateout(reg)` have different meanings (the second guarantees them to use the same register, the first doesn't) when using an explicit register this difference is eliminated (in both cases they are guaranteed to use the same, specified register). And since the second syntax is the one guaranteeing usage of the same register (with no other difference that I can tell), it makes sense for it to be used in the case of the specific register, where this also guaranteed.

Reply:
You make very good points and I have adjusted the RFC accordingly.